### DISCUSSION TITLE
cpp/memory/malloc_free

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#31] [Announcements]
# cpp/memory/malloc_free

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/memory/malloc_free.html

**Comment by junnan-xjtu** [comment]
这篇感觉写的有一点乱了

**Reply by HeatDeathOfUniverse** [reply]
那是相当的乱，感觉可读性很差

**Comment by HeatDeathOfUniverse** [comment]
bins[64,127)称为large_bins，用于维护>512字节的内存块，每个元素对应的链表中的chunk大小不同，index越大，链表中chunk的内存大小相差越大，例如: 下标为64的chunk大小介于[512, 512+64)，下标为95的chunk大小介于[2k+1,2k+512)。同一条链表上的chunk，按照从小到大的顺序排列。
这段表述有点模糊可以看看按照https://ciphersaw.me/ctf-wiki/pwn/linux/heap/heap_structure/#large-bin 这里的解释，为什么 index 越大，内存大小相差越大。然后下标 95 的 chunk 大小范围应该是[2496, 2560)？

**Comment by huangjiawei128** [comment]
一个比较好的malloc和free的简易实现：https://www.bilibili.com/video/BV1Dy4y1579m/?spm_id_from=333.788.recommend_more_video.-1&vd_source=b257fee9bc4fee2c8e3c32a75effdb7f

**Comment by fangtaoysu** [comment]
其中代码段主要存放进程的可执行二进制代码，字符串字面值和只读变量。
这块准确说：只读变量如字符串字面值、const全局变量等在常量区，也就是`.rodata`

**Reply by fangtaoysu** [reply]
其中代码段主要存放进程的可执行二进制代码，字符串字面值和只读变量。

数据段存放已经初始化且初始值非0的全局变量和局部静态变量。

bss段则存放未初始化或初始值为0的全局变量和局部静态变量。
这一整块有些许问题，应该是：
`.text`存放程序指令、函数代码，`.rodata`存放字符串常量、const全局常量，`.data`存放已初始化的全局/静态变量，`.bss`存放为初始化的全局/静态变量
